File: 2001-3000/2518-number-of-great-partitions/2518-number-of-great-partitions.cpp
class Solution {
public:
    int countPartitions(vector<int>& nums, int k) {
        const int MOD = 1e9 + 7;

        long long total = 0;
        for (int x : nums) total += x;

        if (total < 2LL * k) return 0;

        int n = nums.size();

        vector<long long> dp(k, 0);
        dp[0] = 1;

        for (int x : nums) {
            for (int s = k - 1; s >= x; --s) {
                dp[s] = (dp[s] + dp[s - x]) % MOD;
            }
        }

        long long bad = 0;
        for (int s = 0; s < k; ++s) {
            bad = (bad + dp[s]) % MOD;
        }

        long long totalPartitions = 1;
        for (int i = 0; i < n; ++i) {
            totalPartitions = (totalPartitions * 2) % MOD;
        }

        long long ans = (totalPartitions - 2 * bad) % MOD;
        if (ans < 0) ans += MOD;

        return (int)ans;
    }
};

File: 2001-3000/2518-number-of-great-partitions/2518-number-of-great-partitions.java
class Solution {
    public int countPartitions(int[] nums, int k) {
        int n = nums.length, MOD = 1_000_000_007;

        long totalSum = 0;
        for (int num : nums) {
            totalSum += num;
        }

        // base check
        if (totalSum < (2 * k)) {
            return 0;
        }

        int[] dp = new int[k];
        dp[0] = 1;

        for (int num : nums) {
            for (int i = k - 1; i >= num; i--) {
                dp[i] = (dp[i] + dp[i - num]) % MOD;
            }    
        }
        
        long totalPartitions = 1;
        for (int i = 0; i < n; i++) {
            totalPartitions = (totalPartitions * 2) % MOD;
        }

        long totalBadWays = 0;
        for (int i = 0; i < k; i++) {
            totalBadWays = (totalBadWays + dp[i]) % MOD;
        }

        long totalBadPartitions = (totalBadWays * 2) % MOD;

        long greatPartitions = (totalPartitions - totalBadPartitions + MOD) % MOD;

        return (int) greatPartitions;
    }
}